torchaudio/functional.py

```diff
@@ -1433,6 +1433,152 @@ def _generate_wave_table(
+        regen (float): desired regen(feeback gain) in dB
+            Allowed range of values are -95 to 95
+        width (float):  desired width(delay gain) in dB
+            Allowed range of values are 0 to 100
+        speed (float):  modulation speed in Hz
+            Allowed range of values are 0.1 to 10
+        phase (float):  percentage phase-shift for multi-channel
+            Allowed range of values are 0 to 100
+        sinusoidal (bool):  If ``True``, uses sinusoidal modulation
+            If ``False``, uses triangular modulation
+            (Default: ``True``)
+        linear_interpolation (bool):  If ``True``, uses linear interpolation for delay-line interpolation
+            If ``False``, uses Quadratic interpolation
+            (Default: ``True``)
+
+    Returns:
+        Tensor: Waveform of dimension of `(...,channels,time)`
+
+    References:
+        http://sox.sourceforge.net/sox.html
+        Scott Lehman, Effects Explained, http://harmony-central.com/Effects/effects-explained.html
+    """
+
+    actual_shape = waveform.shape
+    device, dtype = waveform.device, waveform.dtype
+
+    if actual_shape[-2] > 4:
+        raise ValueError("Max 4 channels allowed")
+
+    # convert to 3D (batch, channels, time)
+    waveform = waveform.view(-1, actual_shape[-2], actual_shape[-1])
+
+    # Scaling
+    feedback_gain = regen / 100
+    delay_gain = width / 100
+    channel_phase = phase / 100
+    delay_min = delay / 1000
+    delay_depth = depth / 1000
+
+    n_channels = waveform.shape[-2]
+
+    if sinusoidal:
+        wave_type = 'SINE'
+    else:
+        wave_type = 'TRIANGLE'
+
+    # Balance output:
+    in_gain = 1. / (1 + delay_gain)
+    delay_gain = delay_gain / (1 + delay_gain)
+
+    # Balance feedback loop:
+    delay_gain = delay_gain * (1 - abs(feedback_gain))
+
+    delay_buf_length = int((delay_min + delay_depth) * sample_rate + 0.5)
+    delay_buf_length = delay_buf_length + 2
+
+    delay_bufs = torch.zeros(waveform.shape[0], n_channels, delay_buf_length, dtype=dtype, device=device)
+    delay_last = torch.zeros(waveform.shape[0], n_channels, dtype=dtype, device=device)
+
+    lfo_length = int(sample_rate / speed)
+
+    lfo = torch.zeros(lfo_length, dtype=dtype, device=device)
+
+    table_min = math.floor(delay_min * sample_rate + 0.5)
+    table_max = delay_buf_length - 2.
+
+    lfo = _generate_wave_table(wave_type=wave_type,
+                               data_type='FLOAT',
+                               table_size=lfo_length,
+                               min=float(table_min),
+                               max=float(table_max),
+                               phase=3 * math.pi / 2)
+
+    output_waveform = torch.zeros_like(waveform, dtype=dtype, device=device)
+
+    delay_buf_pos = 0
+    lfo_pos = 0
+
+    for i in range(waveform.shape[-1]):
+
+        delay_buf_pos = (delay_buf_pos + delay_buf_length - 1) % delay_buf_length
+
+        for c in range(0, n_channels):
```

Comment:
Correct me if I am wrong, it seems that the same operation with different parameters is applied to each channel individually. If operations among channels do not interweave, can we vectorize the operation?

Looking at the test result, the test of this function takes the longest.
https://app.circleci.com/pipelines/github/pytorch/audio/1768/workflows/4a2bae91-b123-4fcd-af13-a0a24c929c1e/jobs/44621

Reply:
@mthrok , I have vectorized the channels loop , But facing issue in torch script test , I am getting below error 
```
RuntimeError: 
E           Arguments for call are not valid.
E           The following variants are available:
E             
E             aten::frac(Tensor self) -> (Tensor):
E             Expected a value of type 'Tensor' for argument 'self' but instead found type 'float'.
E             
E             aten::frac.out(Tensor self, *, Tensor(a!) out) -> (Tensor(a!)):
E             Expected a value of type 'Tensor' for argument 'self' but instead found type 'float'.
E           
E           The original call is:
E             File "/Users/bhargav/audio/torchaudio/functional.py", line 1542
E                   cur_channel_phase = (torch.arange(0, n_channels) * lfo_length * channel_phase + .5).to(torch.int64)
E                   delay = lfo[(lfo_pos + cur_channel_phase) % lfo_length]
E                   frac_delay = torch.frac(delay)
E                                ~~~~~~~~~~ <--- HERE
E                   delay = torch.floor(delay)
E           'flanger' is being compiled since it was called from 'func'
E             File "/Users/bhargav/audio/test/test_torchscript_consistency.py", line 542
E                       phase = 60.
E                       sample_rate = 44100
E                       return F.flanger(tensor, sample_rate, delay, depth, regen, width, speed,
E                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~...  <--- HERE
E                                        phase, sinusoidal=True, linear_interpolation=True)
```

The delay variable is a tensor, But the torch script is saying it got float instead of tensor

Reply:
`delay` is used in function signature and annotated as `float`, so JIT compiler considers it float but you are also re-defining variable `delay` as Tensor.

Change the variable name to something like `delay_tensor = lfo[(lfo_pos + cur_channel_phase) % lfo_length]`, and it should work.

Reply:
@mthrok , Fixed it , Torch script test passed

Reply:
@bhargavkathivarapu 
Thanks. Can you rebase? In #644, I moved the test file.

Also, in Torchscript test and batch consistency test, instead of loading sample value from a file, can you generate input value, with fixed random seed?

Reply:
@mthrok , I have messed up rebasing ,  I am closing this pull request . Made a new pull request  with suggested changes( #651 ) , please check that .